src/parserTest.cpp

```diff
@@ -0,0 +1,365 @@
+    Node* t = new Node("write", "WRITE", "ellipse");
+    match(TokenType::WRITE);
+    t->setChildren({exp(), nullptr});
+    cout << "Exiting writeStmt()" << endl; // Debugging statement
+    return t;
+}
+
+void Parser::createNodesTable(Node* node) {
+    if (node == nullptr) { // Check if node is null
+        if (parse_tree == nullptr) { // Ensure parse_tree is also initialized
+            cerr << "Error: Parse tree is not initialized." << endl;
+            return;
+        }
+        node = parse_tree; // Start from the root of the parse tree
+    }
+
+    // Process the current node only if it hasn't been added to the table
+    if (node->index == -1) {
+        node->index = nodes_table.size(); // Assign a new index
+        nodes_table[node->index] = {node->token_value, node->code_value, node->shape};
+    }
+
+    // Traverse children recursively
+    for (auto child : node->children) {
+        if (child != nullptr) { // Check for null children
+            createNodesTable(child);
+        }
+    }
+
+    // Traverse sibling recursively
+    if (node->sibling != nullptr) {
+        createNodesTable(node->sibling);
+    }
+}
+
+
+void Parser::createEdgesTable(Node* node) {
+    if (node == nullptr) {
+        node = parse_tree;
+    }
+    for (auto child : node->children) {
+        edges_table.emplace_back(node->index, child->index);
+        createEdgesTable(child);
+    }
+    if (node->sibling != nullptr) {
+        edges_table.emplace_back(node->index, node->sibling->index);
+        createEdgesTable(node->sibling);
+    }
+}
+
+void Parser::run() {
+    cout << "Running parser..." << endl; // Debugging statement
+    parse_tree = stmtSequence();
+    createNodesTable(parse_tree);
+    createEdgesTable(parse_tree);
+    cout << "Finished parsing" << endl; // Debugging statement
+}
+
+void Parser::outputTables() {
+    cout << "Nodes table:" << endl;
+    for (const auto& node : nodes_table) {
+        cout << node.first << ": " << node.second[0] << ", " << node.second[1] << ", " << node.second[2] << endl;
+    }
+    cout << "Edges table:" << endl;
+    for (const auto& edge : edges_table) {
+        cout << edge.first << " -> " << edge.second << endl;
+    }
+}
+
+void Parser::clearTables() {
+    nodes_table.clear();
+    edges_table.clear();
+    same_rank_nodes.clear();
+}
+
+void Parser::printParseTree(Node* node, int depth) {
+    if (node == nullptr) {
+        return;
+    }
+    for (int i = 0; i < depth; i++) {
+        cout << "  ";
+    }
+    cout << node->token_value << endl;
+    for (auto child : node->children) {
+        printParseTree(child, depth + 1);
+    }
+    if (node->sibling != nullptr) {
+        printParseTree(node->sibling, depth);
+    }
+}
+
+int main(int argc, char** argv) {
+    // use static data token to test parser
+    const vector<Token> tokens = {
+        {TokenType::READ, "read", 0},
+        {TokenType::IDENTIFIER, "x", 0},
+    };
```

Comment:
Gives a segfault with the following tokens due to null pointer deference at line 300